File: node/wags-media-repository-api/src/models/movie.ts
import { VideoGenre, VideoService } from './system';

export type MovieStatus = {
    movieStatusId: number;
    name: string;
    colorCode: string;
}

export type Movie = {
    movieId: number;
    title: string;
    imdbLink: string;
    posterImageUrl: string;
    dateWatched?: Date | string | null;
    rating: number;
    thoughts: string;
    sortOrder?: number;
    statusId: number;
    status?: MovieStatus;
    genres: VideoGenre[];
    services: VideoService[]
}

export type MovieGenreLink = {
    movieId: number;
    genreId: number;
    genreName: string;
    genreColorCode: string;
}

export type MovieServiceLink = {
    movieId: number;
    serviceId: number;
    serviceName: string;
    serviceColorCode: string;
}

File: dotnet/src/WagsMediaRepository.Domain/Models/Movie.cs
namespace WagsMediaRepository.Domain.Models;

public class Movie
{
    public int MovieId { get; set; }
    
    public int MovieStatusId { get; set; }

    public string Title { get; set; } = string.Empty;

    public string ImdbLink { get; set; } = string.Empty;
    
    public DateTime? DateWatched { get; set; }
    
    public int? SortOrder { get; set; }
    
    public int Rating { get; set; }

    public string Thoughts { get; set; } = string.Empty;
    
    public string PosterImageUrl { get; set; } = string.Empty;

    public MovieStatus Status { get; set; } = new();

    public List<MovieGenre> Genres { get; set; } = [];

    public List<MovieService> Services { get; set; } = [];

    public static Movie FromDto(MovieDto dto) => new()
    {
        MovieId = dto.MovieId,
        MovieStatusId = dto.MovieStatusId,
        Title = dto.Title,
        ImdbLink = dto.ImdbLink,
        DateWatched = dto.DateWatched,
        SortOrder = dto.SortOrder,
        Rating = dto.Rating,
        Thoughts = dto.Thoughts,
        PosterImageUrl = dto.PosterImageUrl,
        Status = MovieStatus.FromDto(dto.MovieStatus),
        Genres = dto.MovieToMovieGenres
            .Select(mg => MovieGenre.FromDto(mg.MovieGenre))
            .ToList(),
        Services = dto.MovieToMovieServices
            .Select(ms => MovieService.FromDto(ms.MovieService))
            .ToList(),
    };
}